### PR TITLE
Fastq.gz support in hisat2

### PR DESCRIPTION
**Pull request by dpryan79**
This PR finishes part of #1128, adding support for gz and bz2 input into hisat2. Additionally, if one selects the "Write unaligned reads..." or "Write aligned reads..." options, then the output type will match the inputs (i.e., fastqsanger.gz input leads to fastqsanger.gz output).

I added a test for gz and bz2 input. I'm hoping to use this sort of structure for the other tools mentioned in #1128, so feedback regarding that would be appreciated.

### CODE REVIEW

File: tools/hisat2/hisat2.xml [hunk 1]
```diff
@@ -28,19 +31,28 @@
...
-            -1 '${input_format.paired.reads.forward}' -2 '${input_format.paired.reads.reverse}'
+        #if str($input_format.paired.paired_selector) != 'single':
+            -1 '${read1}' -2 '${read2}'
             @paired_end_options@
         #else:
-            -U '${input_format.paired.reads}'
+            -U '${read1}'
             #if str( $input_format.paired.unaligned_file ) == "true":
-                --un '$output_unaligned_reads_l'
+                #if $compressed == "GZ":
```
Comment (bgruening) [inline tools/hisat2/hisat2.xml]: `#if $compressed in ["GZ", "BZ2"]:` can save a few lines.

**Comment by bgruening** [comment]
Thanks Devon!

**Comment by dpryan79** [comment]
I guess I just missed the version bump. If that's needed I can make a quick PR for it.

File: tools/hisat2/hisat2.xml (hunk 1 above)
Comment (dpryan79) [inline tools/hisat2/hisat2.xml]: In this case each compression type has a different argument associated with it, so I'm not sure what that'll really save lines.

**Comment by bgruening** [comment]
Mh, now it's merged ... bad @nekrut ;) ... not sure if the version bump now is necessary it's already in the TS. Let's hope the best.
Thanks @dpryan79!

File: tools/hisat2/hisat2.xml (hunk 1 above)
Comment (bgruening) [inline tools/hisat2/hisat2.xml]: I only see one argument `--un-gz`

**Comment by dpryan79** [comment]
@bgruening At least I'll remember to do that for the other PRs now (just updated my fork with the bowtie2 changes).

**Comment by bgruening** [comment]
You rock! Thanks Devon!

File: tools/hisat2/hisat2_macros.xml [hunk 2]
```diff
@@ -55,15 +55,23 @@
...
+            #if $compressed == "GZ":
+                --un-conc-gz '${output_unaligned_reads_l}'
+            #else:
+                --un-conc '${output_unaligned_reads_l}'
+            #end if
         #end if
         #if str( $input_format.paired.aligned_file ) == "true":
-            --al-conc '${output_aligned_reads_l}'
+            #if $compressed == "GZ":
+                --al-conc '${output_aligned_reads_l}'
```
Comment (nsoranzo) [inline tools/hisat2/hisat2_macros.xml]: --al-conc-gz

File: tools/hisat2/hisat2.xml [hunk 3]
```diff
@@ -28,19 +31,28 @@
...
+                    --un-gz '$output_unaligned_reads_l'
+                #else if $compressed == "BZ2":
+                    --un-gz '$output_unaligned_reads_l'
```
Comment (nsoranzo) [inline tools/hisat2/hisat2.xml]: --un-bz2

File: tools/hisat2/hisat2_macros.xml [hunk 4]
```diff
@@ -77,4 +85,65 @@
...
+            #if $input_format.paired.reads_f.is_of_type("fastq.gz", "fastqsanger.gz"):
+                #set read1 = "input_f.fastq.gz"
+                #set compressed = "GZ"
+            #else if $input_format.paired.reads_f.is_of_type("fastq.bz2", "fastqsanger.bz2"):
+                #set read1 = "input_f.fastq.bz2"
+                #set compressed = "BZ2"
+            #else:
+                #set read1 = "input_f.fastq"
+            #end if
+            ln -f -s "${input_format.paired.reads_f}" "${read1}" &&
```
Comment (nsoranzo) [inline tools/hisat2/hisat2_macros.xml]: Single quotes around `${input_format.paired.reads_f}` (no quotes needed around `$read1`)

File: tools/hisat2/hisat2_macros.xml [hunk 5]
```diff
@@ -77,4 +85,65 @@
...
+            #if $input_format.paired.reads_r.is_of_type("fastq.gz", "fastqsanger.gz"):
+                #set read2 = "input_r.fastq.gz"
+                #set compressed = "GZ"
+            #else if $input_format.paired.reads_r.is_of_type("fastq.bz2", "fastqsanger.bz2"):
+                #set read2 = "input_r.fastq.bz2"
+                #set compressed = "BZ2"
+            #else:
+                #set read2 = "input_r.fastq"
+            #end if
+            ln -f -s "${input_format.paired.reads_r}" "${read2}" &&
```
Comment (nsoranzo) [inline tools/hisat2/hisat2_macros.xml]: As above.

File: tools/hisat2/hisat2_macros.xml [hunk 6]
```diff
@@ -77,4 +85,65 @@
...
+            #if $input_format.paired.reads.forward.is_of_type("fastq.gz", "fastqsanger.gz"):
+                #set read1 = "input_f.fastq.gz"
+                #set compressed = "GZ"
+            #else if $input_format.paired.reads.forward.is_of_type("fastq.bz2", "fastqsanger.bz2"):
+                #set read1 = "input_f.fastq.bz2"
+                #set compressed = "BZ2"
+            #else:
+                #set read1 = "input_f.fastq"
+            #end if
+            ln -s "${input_format.paired.reads.forward}" "${read1}" &&
```
Comment (nsoranzo) [inline tools/hisat2/hisat2_macros.xml]: As above.

File: tools/hisat2/hisat2_macros.xml [hunk 7]
```diff
@@ -77,4 +85,65 @@
...
+            #if $input_format.paired.reads.reverse.is_of_type("fastq.gz", "fastqsanger.gz"):
+                #set read2 = "input_r.fastq.gz"
+                #set compressed = "GZ"
+            #else if $input_format.paired.reads.reverse.is_of_type("fastq.bz2", "fastqsanger.bz2"):
+                #set read2 = "input_r.fastq.bz2"
+                #set compressed = "BZ2"
+            #else:
+                #set read2 = "input_r.fastq"
+            #end if
+            ln -s "${input_format.paired.reads.reverse}" "${read2}" &&
```
Comment (nsoranzo) [inline tools/hisat2/hisat2_macros.xml]: As above.

File: tools/hisat2/hisat2_macros.xml [hunk 8]
```diff
@@ -77,4 +85,65 @@
...
+            #if $input_format.paired.reads.is_of_type("fastq.gz", "fastqsanger.gz"):
+                #set read1 = "input_f.fastq.gz"
+                #set compressed = "GZ"
+            #else if $input_format.paired.reads.is_of_type("fastq.bz2", "fastqsanger.bz2"):
+                #set read1 = "input_f.fastq.bz2"
+                #set compressed = "BZ2"
+            #else:
+                #set read1 = "input_f.fastq"
+            #end if
+            ln -s "${input_format.paired.reads}" "${read1}" &&
```
Comment (nsoranzo) [inline tools/hisat2/hisat2_macros.xml]: As above.

File: tools/hisat2/hisat2_macros.xml [hunk 9]
```diff
@@ -55,15 +55,23 @@
...
-        <param name="aligned_file" type="boolean" truevalue="true" falsevalue="false" checked="False" label="Write aligned reads (in fastq format) to separate file(s)" argument="--al-conc" />
+        <param name="unaligned_file" type="boolean" truevalue="true" falsevalue="false" checked="False" label="Write unaligned reads (in fastq format) to separate file(s)" argument="--un-conc / --un-conc-gz" />
+        <param name="aligned_file" type="boolean" truevalue="true" falsevalue="false" checked="False" label="Write aligned reads (in fastq format) to separate file(s)" argument="--al-conc / --al-conc-gz" />
     </xml>
     <token name="@paired_end_options@">
         #if str( $input_format.paired.unaligned_file ) == "true":
-            --un-conc '${output_unaligned_reads_l}'
+            #if $compressed == "GZ":
+                --un-conc-gz '${output_unaligned_reads_l}'
+            #else:
```
Comment (nsoranzo) [inline tools/hisat2/hisat2_macros.xml]: BZ2?

File: tools/hisat2/hisat2_macros.xml [hunk 10]
```diff
@@ -55,15 +55,23 @@
...
+                --un-conc '${output_unaligned_reads_l}'
+            #end if
         #end if
         #if str( $input_format.paired.aligned_file ) == "true":
-            --al-conc '${output_aligned_reads_l}'
+            #if $compressed == "GZ":
+                --al-conc '${output_aligned_reads_l}'
+            #else:
```
Comment (nsoranzo) [inline tools/hisat2/hisat2_macros.xml]: BZ2?

File: tools/hisat2/hisat2.xml (hunk 3 above)
Comment (dpryan79) [inline tools/hisat2/hisat2.xml]: Thanks @nsoranzo, I'll prepare a new PR with everything you've noticed.

**Comment by nekrut** [comment]
Oops - sorry

**Pull request merged**